lighthouse-core/gather/gatherers/form-elements.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * @license Copyright 2019 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Gatherer = require('./gatherer.js');
+const pageFunctions = require('../../lib/page-functions.js');
+
+/* eslint-env browser, node */
+
+/**
+ *  @param {HTMLElement}
+ *  @return {String}
+ */
+/* istanbul ignore next */
+function getParentForm(node) {
+  if (node == undefined){
+    return [undefined, false];
+  }
+  if (node.nodeName == 'BODY'){
+    return [getNodePath(node), false];
+  };
+  if (node.nodeName == 'FORM'){
+    if (node.id && node.id != ""){
+      return [node.id, true];
+    }
+    if (node.name && node.name != ""){
+      return [node.name, true];
+    }
+    // @ts-ignore - getNodePath put into scope via stringification
+    return [getNodePath(node), false]; // eslint-disable-line no-undef
+  };
+
+  return getParentForm(node.parentElement);
+}
+
+/**
+ * @return {LH.Artifacts['FormElements']}
+ */
+/* istanbul ignore next */
+function collectFormElements() {
+  // @ts-ignore - put into scope via stringification
+  const inputElements = getElementsInDocument('input'); // eslint-disable-line no-undef
+  const selectElements = getElementsInDocument('select'); // eslint-disable-line no-undef
+  const textareaElements = getElementsInDocument('textarea'); // eslint-disable-line no-undef
+  const labelElements = getElementsInDocument('label'); // eslint-disable-line no-undef
+  const formElements = inputElements.concat(selectElements, textareaElements, labelElements)
+  return formElements.map(/** @param {HTMLElement} node */ (node) => {
```

Comment:
do we need any details about the form element? (like how the `<form autocomplete>` attribute is set?)

i was thinking an alternative way to structure this data is to group all input elements by their parent form.  but the downside is you do need to account for elements lacking a parent form. no strong feeling right now, but it may become more clear with the audit's impl.

Reply:
Yea I wanted to account for formless inputs so thats why I went with the bottom up vs a top down approach

Reply:
> i was thinking an alternative way to structure this data is to group all input elements by their parent form. but the downside is you do need to account for elements lacking a parent form. no strong feeling right now, but it may become more clear with the audit's impl.

I think this would be preferable too ([grouping by forms](https://github.com/GoogleChrome/lighthouse/issues/10450#issuecomment-598866825-permalink:~:text=I'm%20thinking%20it%20should%20be%20an%20array%20of%20forms%2C%20like)). What do you think George?

Reply:
@connorjclark the only problem is that it can't account for the formelss inputs

Reply:
You can have a single entry where there is a null value for `formElementPath` or w.e., and all form-less elements go there.

Reply:
@connorjclark so collect all form elements, and collect inputs as well? any inputs with no form go under formless?

Reply:
yup